### PR TITLE
std.net.StreamServer.Options: add reuse_port

### LINKED ISSUE #11361: std.net.StreamServer does not support port reuse
### Zig Version

0.9.1

### Steps to Reproduce

1. code `main.zig`: 
```zig
const std = @import("std");
const StreamServer = std.net.StreamServer;
const Address = std.net.Address;

pub fn main() anyerror!void {
    var server = StreamServer.init(.{
        .reuse_address = true,
    });
    defer server.deinit();

    var addr = try Address.parseIp("127.0.0.1", 8080);
    try server.listen(addr);
    
    while (true) {
        //do sth.
        std.time.sleep(1*std.time.ns_per_s);
    }
    std.log.info("ok.", .{});
}
```
2. build
```shell
zig build
```
3. run
```shell
./zig-out/bin/streamServerPortReuse &
./zig-out/bin/streamServerPortReuse
```

### Expected Behavior

Multiple instances can run without error. 

StreamServer can accept reuse_port parameter.

```zig
const std = @import("std");
const StreamServer = std.net.StreamServer;
const Address = std.net.Address;

pub fn main() anyerror!void {
    var server = StreamServer.init(.{
        .reuse_address = true,
        .reuse_port = true,        // FIXME:
    });
    defer server.deinit();

    var addr = try Address.parseIp("127.0.0.1", 8080);
    try server.listen(addr);
    
    while (true) {
        //do sth.
        std.time.sleep(1*std.time.ns_per_s);
    }
    std.log.info("ok.", .{});
}
```

After patching `/opt/homebrew/Cellar/zig/0.9.1/lib/zig/std/net.zig`, it accepts reuse_port parameter and works as expected.

```patch
diff --git a/net.zig b/net.zig
index f564a9d..942a79b 100644
--- a/net.zig
+++ b/net.zig
@@ -1706,6 +1706,7 @@ pub const StreamServer = struct {
     /// Copied from `Options` on `init`.
     kernel_backlog: u31,
     reuse_address: bool,
+    reuse_port: bool,
 
     /// `undefined` until `listen` returns successfully.
     listen_address: Address,
@@ -1720,6 +1721,9 @@ pub const StreamServer = struct {
 
         /// Enable SO.REUSEADDR on the socket.
         reuse_address: bool = false,
+
+        /// Enable SO.REUSEPORT on the socket.
+        reuse_port: bool = false,
     };
 
     /// After this call succeeds, resources have been acquired and must
@@ -1729,6 +1733,7 @@ pub const StreamServer = struct {
             .sockfd = null,
             .kernel_backlog = options.kernel_backlog,
             .reuse_address = options.reuse_address,
+            .reuse_port = options.reuse_port,
             .listen_address = undefined,
         };
     }
@@ -1759,6 +1764,14 @@ pub const StreamServer = struct {
                 &mem.toBytes(@as(c_int, 1)),
             );
         }
+        if (self.reuse_port) {
+            try os.setsockopt(
+                sockfd,
+                os.SOL.SOCKET,
+                os.SO.REUSEPORT,
+                &mem.toBytes(@as(c_int, 1)),
+            );
+        }
 
         var socklen = address.getOsSockLen();
         try os.bind(sockfd, &address.any, socklen);
```

### Actual Behavior

The first instance works, but the second instance fails.

```shell
aka@akadeMacBook-Pro streamServerPortReuse% ./zig-out/bin/streamServerPortReuse &
[1] 15826
aka@akadeMacBook-Pro streamServerPortReuse % 
aka@akadeMacBook-Pro streamServerPortReuse % ./zig-out/bin/streamServerPortReuse  
error: AddressInUse
/opt/homebrew/Cellar/zig/0.9.1/lib/zig/std/os.zig:3135:27: 0x10291c51b in std.os.bind (streamServerPortReuse)
            .ADDRINUSE => return error.AddressInUse,
                          ^
/opt/homebrew/Cellar/zig/0.9.1/lib/zig/std/net.zig:1764:9: 0x10291958b in std.net.StreamServer.listen (streamServerPortReuse)
        try os.bind(sockfd, &address.any, socklen);
        ^
/Users/aka/github/streamServerPortReuse/src/main.zig:13:5: 0x1029190a3 in main (streamServerPortReuse)
    try server.listen(addr);
    ^

```

### PR DESCRIPTION
**Pull request by arkie**
Applies the patch from #11361 and adds a test for re-binding socket.

Closes #11361

### CODE REVIEW

**Comment by arkie** [comment]
It looks like this has passed the two `Cancelled` Windows tests separately without change. Can this be merged?

**Review by Vexu — APPROVED** [review]

**Comment by arkie** [comment]
> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

Should I keep merging against `master` for test re-runs or just wait re: merge? Thanks!

File: lib/std/net.zig [hunk 1]
```diff
@@ -1920,6 +1925,15 @@ pub const StreamServer = struct {
                 &mem.toBytes(@as(c_int, 1)),
             );
         }
+        // TODO: Investigate if Windows needs an analogue.
```
Comment (DraagrenKirneh) [inline lib/std/net.zig:1928]: Windows do not have "REUSEPORT", instead it has "REUSEADDR". 
There is a good writeup about the subtle differences in this SO post -> https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ

Windows doc on REUSEADDR: https://learn.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse

**Comment by arkie** [comment]
@Vexu Mind merging? Thanks!

**Pull request merged**